File: stock_alerts.go
// stock_alerts.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
	"github.com/google/uuid"
)

type Alert struct {
	ID        string  `json:"id"`
	Symbol    string  `json:"symbol"`
	Threshold float64 `json:"threshold"`
	Above     bool    `json:"above"`
	Triggered bool    `json:"triggered"`
}

type App struct {
	Alerts []Alert `json:"alerts"`
	File   string
}

func NewApp(file string) *App {
	a := &App{File: file}
	a.load()
	return a
}

func (a *App) load() {
	data, err := os.ReadFile(a.File)
	if err != nil {
		return
	}
	json.Unmarshal(data, a)
}

func (a *App) save() {
	data, _ := json.MarshalIndent(a, "", "  ")
	os.WriteFile(a.File, data, 0644)
}

func (a *App) getAlert(symbol string) *Alert {
	for i := range a.Alerts {
		if a.Alerts[i].Symbol == strings.ToUpper(symbol) {
			return &a.Alerts[i]
		}
	}
	return nil
}

func (a *App) add(symbol string, threshold float64, above bool) {
	symbol = strings.ToUpper(symbol)
	// Remove existing
	a.Alerts = []Alert{}
	for _, al := range a.Alerts {
		if al.Symbol != symbol {
			a.Alerts = append(a.Alerts, al)
		}
	}
	alert := Alert{
		ID:        uuid.New().String()[:8],
		Symbol:    symbol,
		Threshold: threshold,
		Above:     above,
		Triggered: false,
	}
	a.Alerts = append(a.Alerts, alert)
	a.save()
	fmt.Printf("✅ Added alert: %s %s %.2f\n", symbol, func() string {
		if above { return ">" } else { return "<" }
	}(), threshold)
}

func (a *App) remove(symbol string) {
	symbol = strings.ToUpper(symbol)
	newAlerts := []Alert{}
	for _, al := range a.Alerts {
		if al.Symbol != symbol {
			newAlerts = append(newAlerts, al)
		}
	}
	a.Alerts = newAlerts
	a.save()
	fmt.Printf("✅ Removed alert for %s\n", symbol)
}

func (a *App) list() {
	if len(a.Alerts) == 0 {
		fmt.Println("No alerts.")
		return
	}
	fmt.Println("\n📋 Current alerts:")
	for _, al := range a.Alerts {
		status := "⏳"
		if al.Triggered {
			status = "🔔"
		}
		sym := al.Symbol
		op := ">"
		if !al.Above {
			op = "<"
		}
		fmt.Printf("  %s: %s %.2f %s\n", sym, op, al.Threshold, status)
	}
}

func fetchPrice(symbol string) (float64, error) {
	url := fmt.Sprintf("https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=%s&apikey=demo", symbol)
	resp, err := http.Get(url)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return 0, err
	}
	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return 0, err
	}
	quote, ok := data["Global Quote"].(map[string]interface{})
	if !ok {
		return 0, fmt.Errorf("no quote data")
	}
	priceStr, ok := quote["05. price"].(string)
	if !ok {
		return 0, fmt.Errorf("price not found")
	}
	return strconv.ParseFloat(priceStr, 64)
}

func (a *App) checkAlerts() {
	for i := range a.Alerts {
		price, err := fetchPrice(a.Alerts[i].Symbol)
		if err != nil {
			fmt.Printf("Error fetching %s: %v\n", a.Alerts[i].Symbol, err)
			continue
		}
		if a.Alerts[i].Above && price > a.Alerts[i].Threshold {
			if !a.Alerts[i].Triggered {
				a.Alerts[i].Triggered = true
				a.save()
				fmt.Printf("\n🔔 ALERT: %s price is %.2f (above %.2f)!\n", a.Alerts[i].Symbol, price, a.Alerts[i].Threshold)
			}
		} else if !a.Alerts[i].Above && price < a.Alerts[i].Threshold {
			if !a.Alerts[i].Triggered {
				a.Alerts[i].Triggered = true
				a.save()
				fmt.Printf("\n🔔 ALERT: %s price is %.2f (below %.2f)!\n", a.Alerts[i].Symbol, price, a.Alerts[i].Threshold)
			}
		}
	}
}

func (a *App) watch(interval int) {
	fmt.Printf("📈 Monitoring alerts every %ds (press Ctrl+C to stop)\n", interval)
	ticker := time.NewTicker(time.Duration(interval) * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			a.checkAlerts()
		}
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: stock_alerts <command> [options]")
		return
	}
	app := NewApp("alerts.json")
	cmd := os.Args[1]

	switch cmd {
	case "add":
		addCmd := flag.NewFlagSet("add", flag.ExitOnError)
		symbol := addCmd.String("symbol", "", "")
		threshold := addCmd.Float64("threshold", 0, "")
		above := addCmd.Bool("above", true, "")
		below := addCmd.Bool("below", false, "")
		addCmd.Parse(os.Args[2:])
		if *symbol == "" && len(addCmd.Args()) > 0 {
			*symbol = addCmd.Args()[0]
		}
		if *threshold == 0 && len(addCmd.Args()) > 1 {
			*threshold, _ = strconv.ParseFloat(addCmd.Args()[1], 64)
		}
		if *threshold == 0 {
			fmt.Println("threshold required")
			return
		}
		aboveFlag := *above
		if *below {
			aboveFlag = false
		}
		app.add(*symbol, *threshold, aboveFlag)

	case "remove":
		if len(os.Args) < 3 {
			fmt.Println("remove <symbol>")
			return
		}
		app.remove(os.Args[2])

	case "list":
		app.list()

	case "watch":
		watchCmd := flag.NewFlagSet("watch", flag.ExitOnError)
		interval := watchCmd.Int("interval", 60, "")
		watchCmd.Parse(os.Args[2:])
		app.watch(*interval)

	default:
		fmt.Println("Unknown command. Use add, remove, list, watch.")
	}
}

File: StockAlerts.cs
// StockAlerts.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

class Alert
{
    [JsonPropertyName("id")] public string Id { get; set; }
    [JsonPropertyName("symbol")] public string Symbol { get; set; }
    [JsonPropertyName("threshold")] public double Threshold { get; set; }
    [JsonPropertyName("above")] public bool Above { get; set; }
    [JsonPropertyName("triggered")] public bool Triggered { get; set; }

    public Alert() { }
    public Alert(string symbol, double threshold, bool above)
    {
        Id = Guid.NewGuid().ToString().Substring(0,8);
        Symbol = symbol.ToUpper();
        Threshold = threshold;
        Above = above;
        Triggered = false;
    }
}

class App
{
    private List<Alert> alerts = new List<Alert>();
    private readonly string configFile = "alerts.json";
    private readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };
    private static readonly HttpClient client = new HttpClient();
    private const string ApiKey = "demo";
    private const string BaseUrl = "https://www.alphavantage.co/query";

    public App() => Load();

    private void Load()
    {
        if (!File.Exists(configFile)) return;
        string json = File.ReadAllText(configFile);
        alerts = JsonSerializer.Deserialize<List<Alert>>(json) ?? new List<Alert>();
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(alerts, options);
        File.WriteAllText(configFile, json);
    }

    private Alert GetAlert(string symbol)
        => alerts.FirstOrDefault(a => a.Symbol.Equals(symbol, StringComparison.OrdinalIgnoreCase));

    public void Add(string symbol, double threshold, bool above)
    {
        var existing = GetAlert(symbol);
        if (existing != null) alerts.Remove(existing);
        alerts.Add(new Alert(symbol, threshold, above));
        Save();
        Console.WriteLine($"✅ Added alert: {symbol} {(above ? ">" : "<")} {threshold:F2}");
    }

    public void Remove(string symbol)
    {
        var existing = GetAlert(symbol);
        if (existing != null)
        {
            alerts.Remove(existing);
            Save();
            Console.WriteLine($"✅ Removed alert for {symbol.ToUpper()}");
        }
        else
        {
            Console.WriteLine($"No alert for {symbol.ToUpper()}");
        }
    }

    public void List()
    {
        if (!alerts.Any())
        {
            Console.WriteLine("No alerts.");
            return;
        }
        Console.WriteLine("\n📋 Current alerts:");
        foreach (var a in alerts)
        {
            string status = a.Triggered ? "🔔" : "⏳";
            Console.WriteLine($"  {a.Symbol}: {(a.Above ? ">" : "<")} {a.Threshold:F2} {status}");
        }
    }

    private async Task<double?> FetchPrice(string symbol)
    {
        try
        {
            string url = $"{BaseUrl}?function=GLOBAL_QUOTE&symbol={symbol}&apikey={ApiKey}";
            var response = await client.GetAsync(url);
            response.EnsureSuccessStatusCode();
            string json = await response.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(json);
            var quote = doc.RootElement.GetProperty("Global Quote");
            return quote.GetProperty("05. price").GetDouble();
        }
        catch
        {
            Console.WriteLine($"Error fetching {symbol}");
            return null;
        }
    }

    public async Task CheckAlerts()
    {
        foreach (var a in alerts)
        {
            var price = await FetchPrice(a.Symbol);
            if (!price.HasValue) continue;
            bool triggered = a.Above ? price > a.Threshold : price < a.Threshold;
            if (triggered && !a.Triggered)
            {
                a.Triggered = true;
                string direction = a.Above ? "above" : "below";
                Console.WriteLine($"\n🔔 ALERT: {a.Symbol} price is {price:F2} ({direction} {a.Threshold:F2})!");
                Save();
            }
        }
    }

    public async Task Watch(int interval)
    {
        Console.WriteLine($"📈 Monitoring alerts every {interval}s (press Ctrl+C to stop)");
        while (true)
        {
            await CheckAlerts();
            await Task.Delay(interval * 1000);
        }
    }

    static async Task Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: StockAlerts <command> [options]");
            return;
        }
        var app = new App();
        var cmd = args[0];

        switch (cmd)
        {
            case "add":
                if (args.Length < 3) { Console.WriteLine("add <symbol> <threshold> [--above|--below]"); return; }
                string symbol = args[1];
                double threshold = double.Parse(args[2]);
                bool above = true;
                for (int i=3; i<args.Length; i++)
                    if (args[i] == "--below") above = false;
                app.Add(symbol, threshold, above);
                break;
            case "remove":
                if (args.Length < 2) { Console.WriteLine("remove <symbol>"); return; }
                app.Remove(args[1]);
                break;
            case "list":
                app.List();
                break;
            case "watch":
                int interval = 60;
                for (int i=1; i<args.Length; i++)
                    if (args[i] == "--interval" && i+1 < args.Length)
                        interval = int.Parse(args[++i]);
                await app.Watch(interval);
                break;
            default:
                Console.WriteLine("Unknown command.");
                break;
        }
    }
}
